### DISCUSSION TITLE
Are there any best practices for UA client reconnection?

### DISCUSSION THREAD

**Discussion by ryangu18** [FreeOpcUa/python-opcua#1462] [Q&A]
I use the client code to connect to the opc ua server and read the data, but when the network is not good, the connection will be interrupted, resulting in abnormal data reading

```
from opcua import Client
client = Client("opc.tcp://192.168.32.239:49320")
client.connect()
while True:  # read data loop
   try:
       """
       read data from UA server
       """
   except Exception as e1:
       	try:
		client.disconnect()  # Its execution is invalid when the network is down
	except:
		pass
	client.connect()
        sleep(10)

```

I modified the code and added the reconnect command, which looks good, but I found that it can cause the number of connections ( i=2277, "CurrentSessionCount") to the server to rise rapidly (especially when the network is interrupted frequently)
Is there a way to reconnect to the server without adding connections?
Or clear invalid connections before connecting? I don't know if this is a function of Client, I tried using client.disconnect()
But it doesn't work, it doesn't close the connection, or reduce the connection, because the network was outage, this client.disconnect() command is invalid.

thank you first

**Comment by AndreasHeine** [comment]
@ryangu18 yes sure!
https://reference.opcfoundation.org/v104/Core/docs/Part4/6.7/

for simple use cases this works robust:
https://github.com/AndreasHeine/opcua-sub-to-mqtt/blob/0fedce03ba0c5719badb4c2e751ac41882540561/opcua_client_mqtt_publisher.py#L111

**Reply by torjusn** [reply]
@AndreasHeine Thank you Andreas. 

Your answer seems like a great solution to the popular question on how to better handle connection loss compared to a `while True`:
https://github.com/FreeOpcUa/opcua-asyncio/discussions/623
https://github.com/FreeOpcUa/opcua-asyncio/issues/139

I think this example (or a barebones without mqtt) added to examples or referenced in the README of python-opcua could benefit many.

**Reply by AndreasHeine** [reply]
the issue with that, is it depends very much on your application... there are other ways to solve this!
in the past i often had to relay data to different protocols or systems so that is where the example came from (in a simplified version!)
